### PR TITLE
EKG counter for tips served

### PR DESCRIPTION
**Pull request by karknu**
Add an EKG counter, cardano.node.metrics.served.block.latest.count, which
can be used to track how usefull the node is to its down stream peers.

This counter can be used as follows `rate(cardano_node_metrics_served_block_latest_count_int[1h])/rate(cardano_node_metrics_blockNum_int[1h])` . Which gives a value close to 0 for unpopular node. And a value around 1 or more for a very useful node.

### CODE REVIEW

File: cardano-node/src/Cardano/Tracing/Tracers.hs [hunk 1]
```diff
@@ -621,6 +616,54 @@ mkConsensusTracers mbEKGDirect trSel verb tr nodeKern fStats = do
...
+traceBlockFetchServerMetrics (Just ekgDirect) tBlocksServed tLocalUp tMaxSlotNo tracer = Tracer bsTracer
+
+  where
+    bsTracer :: TraceBlockFetchServerEvent blk -> IO ()
+    bsTracer e@(TraceBlockFetchServerSendBlock p) = do
+      traceWith tracer e
+      let mSlotNo = case pointSlot p of
+                         Origin -> Nothing
+                         At slotNo -> Just slotNo
+
```
Comment (coot) [inline cardano-node/src/Cardano/Tracing/Tracers.hs:637]: It doesn't seem necessary to transform one maybe type into another one.  A few lines below we pattern match on `mSlotNo`, equally well we can just pattern match on `pointSlot p`.

File: cardano-node/src/Cardano/Tracing/Tracers.hs [hunk 2]
```diff
@@ -621,6 +616,54 @@ mkConsensusTracers mbEKGDirect trSel verb tr nodeKern fStats = do
...
+          served <- STM.modifyReadTVar' tBlocksServed (+1)
+          maxSlotNo <- STM.readTVar tMaxSlotNo
+          case mSlotNo of
+               Nothing ->
+                   return (served, Nothing)
+               Just slotNo ->
+                   case compare maxSlotNo slotNo of
+                        LT -> do
+                            STM.writeTVar tMaxSlotNo slotNo
+                            lu <- STM.modifyReadTVar' tLocalUp (+1)
```
Comment (coot) [inline cardano-node/src/Cardano/Tracing/Tracers.hs:648]: What is `modifyReadTVar'`?

Is it equialvent to:
```
STM.modifyTVar' tLocalUP (+1) >> STM.readTVar tLocalUP
```

File: cardano-node/src/Cardano/Tracing/Tracers.hs [hunk 3]
```diff
@@ -621,6 +616,54 @@ mkConsensusTracers mbEKGDirect trSel verb tr nodeKern fStats = do
...
+                            return (served, Just lu)
+                        GT -> do
+                            return (served, Nothing)
+                        EQ -> do
+                            lu <- STM.modifyReadTVar' tLocalUp (+1)
+                            return (served, Just lu)
+      sendEKGDirectInt ekgDirect
```
Comment (coot) [inline cardano-node/src/Cardano/Tracing/Tracers.hs:655]: Please add an empty line above.

File: cardano-node/src/Cardano/Tracing/Tracers.hs [hunk 4]
```diff
@@ -621,6 +616,54 @@ mkConsensusTracers mbEKGDirect trSel verb tr nodeKern fStats = do
...
+  -> STM.TVar Int64
+  -> STM.TVar Int64
+  -> STM.TVar SlotNo
+  -> Tracer IO (TraceBlockFetchServerEvent blk)
+  -> Tracer IO (TraceBlockFetchServerEvent blk)
+traceBlockFetchServerMetrics Nothing _ _ _ tracer = tracer
+traceBlockFetchServerMetrics (Just ekgDirect) tBlocksServed tLocalUp tMaxSlotNo tracer = Tracer bsTracer
+
+  where
+    bsTracer :: TraceBlockFetchServerEvent blk -> IO ()
```
Comment (coot) [inline cardano-node/src/Cardano/Tracing/Tracers.hs:631]: This adds the local upstreaminess metric which tries to answer the question: how many times we serve a block which slot is either equal or ahead of last served block.

I would like to see a description how this related to the upstreaminess metric that is used to choose peers.

File: cardano-node/src/Cardano/Tracing/Tracers.hs (hunk 2 above)
Comment (karknu) [reply cardano-node/src/Cardano/Tracing/Tracers.hs:648]: https://github.com/input-output-hk/cardano-node/blob/aa7d41199227029cf80b1d5672c5401d1c73f57f/cardano-node/src/Cardano/Node/STM.hs#L24-L30

File: cardano-node/src/Cardano/Tracing/Tracers.hs (hunk 4 above)
Comment (karknu) [reply cardano-node/src/Cardano/Tracing/Tracers.hs:631]: I don't think we should mention that at all. This is a useful metric on its own, regardless of if that is how down stream peers may judge the node in the future.

**Comment by dorin100** [comment]
more useful details for this new metric:
```
A value of 0 means that no one is fetching blocks from your relay.
A value of 1 means that for every new block forged someone is fetching it.
Above 2 means that for every block forged two nodes are fetching that block from you
```

**Pull request merged**